Add method to dependencies and topologies in compact input

srmurguia opened the pull request:
- Add a method field in FromToDependencyConfig
- Cache temporal blueprints that simplifies code when generating
bluprints and allows to have multiple dependency methods (this was a
bug).

srmurguia committed:
Add method to dependencies and topologies in compact input

- Add a method field in FromToDependencyConfig
- Cache temporal blueprints that simplifies code when generating
bluprints and allows to have multiple dependency methods (this was a
bug).

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/models/Topologies.kt:
```diff
@@ -26,6 +27,7 @@ import java.util.*
 enum class Topologies {
     FULL {
         override fun generateDependencies(parameters: Map<String, String>, configPOJO: ConfigPOJO): List<FromToDependencyConfig> {
+            val method = getMethod(parameters)
```
Not for this PR, but we probably need to  find a way to refactor these topologies a bit, because they look very similar. Although I'm not sure if that is possible.

srmurguia replied:
Every time I modify this file I think the same, there should be a way to have all common initialization in a single place. I think it can be done by moving all common variables to be private members and have a method that initialize them. The structure of generateDependencies will be like this:

- Call common initialization
- Initialize specific parameters
- Check for non used parameters, to prevent typo errors from input (this is not done yet)
- Generate dependencies

Once this PR is merged I will work on having that.

NikitaKozlov replied:
It would be awesome if we can bring topologies to the Full format. But that'll require some work to be done :)

srmurguia replied:
PR #107 implements the ideas from my previous comment.

I also want to make them work with the Full format, right now they depend on the names of the modules, that can be easily solved by passing this list instead of having it hard coded as it is now. I can do this as a first step to have them in the Full format. One benefit of having topologies of doing that is that it will be possible to apply topologies only to subsets of modules.

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/models/Topologies.kt:
```diff
@@ -213,6 +222,8 @@ enum class Topologies {
 
     protected fun getWideness(parameters: Map<String, String>, default: Float = 0.5f) = parameters["wideness"]?.toFloat() ?: default
 
+    protected fun getMethod(parameters: Map<String, String>) = parameters.getOrDefault("method", DEFAULT_DEPENDENCY_METHOD)
```
There is probably a bug here, the same for the "full" format. Default dependency method is "implementation", but pure java/kotlin modules do not have it. I suggest you to make `FromToDependencyConfig.method` an optional parameter, as it is in "DependencyConfig" and then the resolution (so as a bug) will stay in a single place.

srmurguia replied:
Changed to use String? and making the default null, now when converted to DependencyConfig, this value is passed to it and hence processed in a single place.

NikitaKozlov commented on src/test/kotlin/com/google/androidstudiopoet/converters/ConfigPojoToAndroidModuleConfigConverterTest.kt:
```diff
@@ -105,7 +106,7 @@ class ConfigPojoToAndroidModuleConfigConverterTest {
         assertOn(androidModuleConfig) {
             hasLaunchActivity.assertTrue()
-            dependencies!!.assertEquals(listOf(DependencyConfig.ModuleDependencyConfig(ANDROID_MODULE_NAME_1)) + PURE_MODULE_DEPENDENCY_LIST)
+            dependencies!!.assertEquals(listOf(DependencyConfig.ModuleDependencyConfig(ANDROID_MODULE_NAME_1, DEFAULT_DEPENDENCY_METHOD)) + PURE_MODULE_DEPENDENCY_LIST)
```
Can you please write a test that checks proper propagation of `method` from `FromToDependencyConfig` to `ModuleDependencyConfig`? Same for `ConfigPojoToModuleConfigConverter`

srmurguia replied:
`convert passes correct values to result AndroidModuleConfig` already does that for AndroidModule, just changed the value to not use DEFAULT_DEPENDENCY_METHOD

srmurguia replied:
Also modified `convert passes correct values to result ModuleConfig` to also test it on ConfigPojoToModuleConfigConverter

srmurguia committed:
Change FromToDependencyConfig#method to String?

instead of assigning a default method, this value is also passed to
DependencyConfig, causing all defaults to be processed by toDepenedencyMethod.
